subsys/bluetooth/host/hci_core.c

```diff
@@ -853,41 +853,62 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 	}
 #endif
 
-	if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
-		if (evt->status) {
+	if (evt->status) {
+		/*
+		 * If there was an error we are only interested in pending
+		 * connection. There is no need to check ID address as
+		 * only one connection can be in that state.
+		 *
+		 * Depending on error code address might not be valid anyway.
+		 */
+		conn = find_pending_connect(NULL);
+		if (!conn) {
+			return;
+		}
+
+		conn->err = evt->status;
+
+		if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
 			/*
-			 * if there was an error we are only interested in pending
-			 * connection so there is no need to check ID address as
-			 * only one connection can be in that state
-			 *
-			 * Depending on error code address might not be valid anyway.
+			 * Handle advertising timeout after high duty directed
+			 * advertising.
 			 */
-			conn = find_pending_connect(NULL);
-			if (!conn) {
-				return;
+			if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
+				atomic_clear_bit(bt_dev.flags,
+						 BT_DEV_ADVERTISING);
+				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+				goto done;
 			}
+		}
 
-			conn->err = evt->status;
+		if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
+			/*
+			 * Handle cancellation of outgoing connection attempt.
+			 */
+			if (conn->err == BT_HCI_ERR_UNKNOWN_CONN_ID) {
+				/* We notify before checking autoconnect flag
+				 * as application may choose to change it from
+				 * callback.
+				 */
+				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+				/* Check if device is marked for autoconnect. */
+				if (atomic_test_bit(conn->flags,
+						    BT_CONN_AUTO_CONNECT)) {
+					bt_conn_set_state(conn,
+							  BT_CONN_CONNECT_SCAN);
+				}
 
-		/*
-		 * Handle advertising timeout after high duty directed
-		 * advertising.
-		 */
-		if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
-			atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
+				goto done;
+			}
 		}
 
-			bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+		BT_WARN("Unexpected status 0x%02x", evt->status);
```

Comment:
Won't this now potentially leave a connection in "connecting" state even though we got a connection complete with error status? That isn't much better than an assert, since it may leave the device non-functional. For the central branch I'd have logged the error `if (conn->err != BT_HCI_ERR_UNKNOWN_CONN_ID)` but otherwise done the same steps for the connection to clean up the state.

Reply:
My point is, even if we get a unexpected error, the one thing we do know is that the connection is disconnected, so leaving it in any other state is probably not a good idea. @sjanc btw, where in the core spec does it say that we can't trust the bdaddr if we get a non-zero status?

Reply:
Considering the old code was merge mess having the known issues handled is a huge improvement. 
Personally I would stick with assert (note that original code was using `BT_ASSERT(evt->status == 0)`) especially if no known errors come to your mind.

Anyway I don't feel competent to add any more changes. Please provide a snippet that I should put and please agree among yourselves what should be there. When it is ready and you reached an agreement please call me.

Reply:
@jhedberg iit is described in Vol 2, Part E, 4.5 "Command Error Handling"

If we get other status code we won't be able to determine which connection failed to establish anyway in case both roles are pending (which leads to another issue, we should not use find_pending_connect() but role dependent function for locating pending conn).

So to avoid assert we can report error but host will probably be left in broken state whatever we do...